shared/elements/gaia_grid/js/grid_dragdrop.js

```diff
@@ -489,16 +489,35 @@
         }
 
         // Cancel rearrangement if it would have no effect.
-        if (this.gridView.items[foundIndex].element.classList.
-            contains('invalid-drop')) {
-          if (insertDividerAtTop) {
-            this.container.classList.remove('hover-over-top');
+        var redundantRearrange = false;
+        if (insertDividerAtTop) {
+          if (iconIsDivider) {
+            redundantRearrange = true;
+            for (var i = foundIndex - 1; i >= 0; i--) {
+              if (this.gridView.items[i].detail.type === 'divider') {
+                redundantRearrange = false;
+                break;
+              }
+            }
+          } else {
+            if (this.icon.detail.index === 0 &&
+                this.gridView.items[1].detail.type === 'placeholder') {
+              redundantRearrange = true;
+            }
           }
+        } else {
+          redundantRearrange = this.gridView.items[foundIndex].element.
+            classList.contains('invalid-drop')
+        }
+
+        if (redundantRearrange) {
           if (this.hoverItem) {
```

Comment:
How do you feel about simply removing the ```redundantRearrange``` variable and moving this code into the ```} else {``` statement on line 508? It seems redundant to have an if statement when we could just move the code into the blocks above.